perftest: init v8.5.0 LTS perftest general-mode vs tuning-mode result (#19558)
pingcap/docs: tidb-cloud/v8.5-performance-benchmarking-with-tpcc.md

---

# TiDB Cloud TPC-C Performance Test Report for TiDB v8.5.0

This document provides the TPC-C performance test steps and results for a TiDB Cloud Dedicated cluster with the TiDB version of v8.5.0. This report can also be used as a reference for the performance of TiDB Self-Managed v8.5.0 clusters.

## Test overview

This test aims at showing the TPC-C performance of TiDB v8.5.0 in the Online Transactional Processing (OLTP) scenario.

## Test environment

### TiDB cluster

The test is conducted on a TiDB cluster with the following settings:

- **Cluster type**: [TiDB Cloud Dedicated](/tidb-cloud/select-cluster-tier.md#tidb-cloud-dedicated)
- **Cluster version**: v8.5.0
- **Cloud provider**: AWS (us-west-2)
- **Cluster configuration**:

| Node type | Node size | Node quantity | Node storage |
|:----------|:----------|:----------|:----------|
| TiDB      | 16 vCPU, 32 GiB | 2 | N/A |
| TiKV      | 16 vCPU, 64 GiB | 3 | 1000 GiB |

### Parameter configuration

> **Note:**
>
> For TiDB Cloud, to modify the TiKV parameters of your cluster, you can contact [PingCAP Support](/tidb-cloud/tidb-cloud-support.md) for help.

> **Tuning Mode Parameters and Variables:**
> 
> For the tuning mode general parameters and variables, use the following value:

Reviewer: ```suggestion
The following are configurations used in [test steps](#test-steps) for tuning mode:
```

---
title: TiDB Cloud TPC-C Performance Test Report for TiDB v8.5.0
summary: Introduce the TPC-C performance test results for a TiDB Cloud Dedicated cluster with the TiDB version of v8.5.0.
---

# TiDB Cloud TPC-C Performance Test Report for TiDB v8.5.0

This document provides the TPC-C performance test steps and results for a TiDB Cloud Dedicated cluster with the TiDB version of v8.5.0. This report can also be used as a reference for the performance of TiDB Self-Managed v8.5.0 clusters.

## Test overview

This test aims at showing the TPC-C performance of TiDB v8.5.0 in the Online Transactional Processing (OLTP) scenario.

## Test environment

### TiDB cluster

The test is conducted on a TiDB cluster with the following settings:

> - **Cluster type**: [TiDB Cloud Dedicated](/tidb-cloud/select-cluster-tier.md#tidb-cloud-dedicated)
> - **Cluster version**: v8.5.0
> - **Cloud provider**: AWS (us-west-2)
> - **Cluster configuration**:
> 
> | Node type | Node size | Node quantity | Node storage |
> |:----------|:----------|:----------|:----------|
> | TiDB      | 16 vCPU, 32 GiB | 2 | N/A |
> | TiKV      | 16 vCPU, 64 GiB | 3 | 1000 GiB |

Reviewer: ```suggestion
- Cluster type: [TiDB Cloud Dedicated](/tidb-cloud/select-cluster-tier.md#tidb-cloud-dedicated)
- Cluster version: v8.5.0
- Cloud provider: AWS (us-west-2)
- Cluster configuration:

    | Node type | Node size | Node quantity | Node storage |
    |:----------|:----------|:----------|:----------|
    | TiDB      | 16 vCPU, 32 GiB | 2 | N/A |
    | TiKV      | 16 vCPU, 64 GiB | 3 | 1000 GiB |
```

### Parameter configuration

> **Note:**
>

> > For TiDB Cloud, to modify the TiKV parameters of your cluster, you can contact [PingCAP Support](/tidb-cloud/tidb-cloud-support.md) for help.

Reviewer: ```suggestion
> For TiDB Cloud, to modify the TiDB or TiKV parameters of your cluster, you can contact [PingCAP Support](/tidb-cloud/tidb-cloud-support.md) for help.
```

The test is conducted in the following two modes:

- General mode: uses default configurations

> - Tune mode: uses optimized configurations

Reviewer 2: ```suggestion
- Tuning mode: uses optimized configurations
```
Keep it consistent with:
https://github.com/pingcap/docs/blob/f636063ebb53d2aa9d7efb9810528b0f2e9556ac/tidb-cloud/v8.5-performance-benchmarking-with-tpcc.md?plain=1#L41

The following are configurations used for tuning mode in [test steps](#test-steps):

- TiDB parameters:

    ```yaml
    [performance]
    concurrently-init-stats = true
    force-init-stats = true
    lite-init-stats = false

    [pessimistic-txn]
    max-retry-count = 15

    [tikv-client]
    region-cache-ttl = 1200
    ```

>     - TiKV parameters:

Reviewer: ```suggestion
- TiKV parameters:
```

    ```yaml
    [server]
    concurrent-send-snap-limit = 64
    concurrent-recv-snap-limit = 64
    snap-io-max-bytes-per-sec = "400MB"

    [raftstore]
    store-io-pool-size = 1
    apply-pool-size = 3
    apply-max-batch-size = 512
    store-max-batch-size = 512

    [raft-engine]
    prefill-for-recycle = true

    [rocksdb.titan]
    enabled = true

>     [rocksdb.defaultcf.titan]

Reviewer 2: ```suggestion

    [rocksdb.defaultcf.titan]
```

    min-blob-size = "1KB"
    blob-file-compression = "zstd"

    [storage.flow-control]
    l0-files-threshold = 60

    [resolved-ts]
    enable = false

    [resource-control]
    enabled = false
    ```

>     - TiDB variables:

Reviewer: ```suggestion
- TiDB variables:
```

    ```sql

>     set global tidb_prepared_plan_cache_size=200;
>     set global tidb_enable_non_prepared_plan_cache=on;
>     set global tidb_ignore_prepared_cache_close_stmt=on;
>     set global tidb_enable_inl_join_inner_multi_pattern=on;
>     set global tidb_opt_derive_topn=on;
>     set global tidb_opt_ordering_index_selectivity_threshold=0.0001;
>     set global tidb_runtime_filter_mode=LOCAL;
>     set global tidb_opt_enable_mpp_shared_cte_execution=on;
>     set global tidb_rc_read_check_ts=on;
>     set global tidb_guarantee_linearizability=off;
>     set global tidb_enable_historical_stats=off;
>     set global tidb_analyze_skip_column_types="json,blob,mediumblob,longblob,mediumtext,longtext";
>     set global tidb_opt_prefer_range_scan=on;
>     set global tidb_max_chunk_size=128;
>     set @@global.tidb_enable_stmt_summary=0;
>     set global tidb_enable_stmt_summary=off;
>     set global tidb_enable_collect_execution_info=off;
>     set global tidb_enable_resource_control=off;
>     set global tidb_txn_assertion_level=off;

Reviewer: ```suggestion
    SET global tidb_prepared_plan_cache_size=200;
    SET global tidb_enable_non_prepared_plan_cache=on;
    SET global tidb_ignore_prepared_cache_close_stmt=on;
    SET global tidb_enable_inl_join_inner_multi_pattern=on;
    SET global tidb_opt_derive_topn=on;
    SET global tidb_opt_ordering_index_selectivity_threshold=0.0001;
    SET global tidb_runtime_filter_mode=LOCAL;
    SET global tidb_opt_enable_mpp_shared_cte_execution=on;
    SET global tidb_rc_read_check_ts=on;
    SET global tidb_guarantee_linearizability=off;
    SET global tidb_enable_historical_stats=off;
    SET global tidb_analyze_skip_column_types="json,blob,mediumblob,longblob,mediumtext,longtext";
    SET global tidb_opt_prefer_range_scan=on;
    SET global tidb_max_chunk_size=128;
    SET @@global.tidb_enable_stmt_summary=0;
    SET global tidb_enable_stmt_summary=off;
    SET global tidb_enable_collect_execution_info=off;
    SET global tidb_enable_resource_control=off;
    SET global tidb_txn_assertion_level=off;
```

> - TiDB variables:
> 
>     ```sql
>     SET global tidb_prepared_plan_cache_size=200;
>     SET global tidb_enable_non_prepared_plan_cache=on;
>     SET global tidb_ignore_prepared_cache_close_stmt=on;
>     SET global tidb_enable_inl_join_inner_multi_pattern=on;
>     SET global tidb_opt_derive_topn=on;
>     SET global tidb_opt_ordering_index_selectivity_threshold=0.0001;
>     SET global tidb_runtime_filter_mode=LOCAL;
>     SET global tidb_opt_enable_mpp_shared_cte_execution=on;
>     SET global tidb_rc_read_check_ts=on;
>     SET global tidb_guarantee_linearizability=off;
>     SET global tidb_enable_historical_stats=off;
>     SET global tidb_analyze_skip_column_types="json,blob,mediumblob,longblob,mediumtext,longtext";
>     SET global tidb_opt_prefer_range_scan=on;
>     SET global tidb_max_chunk_size=128;
>     SET @@global.tidb_enable_stmt_summary=0;
>     SET global tidb_enable_stmt_summary=off;
>     SET global tidb_enable_collect_execution_info=off;
>     SET global tidb_enable_resource_control=off;
>     SET global tidb_txn_assertion_level=off;

Reviewer: ````suggestion
- TiDB system variables:

    ```sql
    SET GLOBAL tidb_prepared_plan_cache_size=200;
    SET GLOBAL tidb_enable_non_prepared_plan_cache=on;
    SET GLOBAL tidb_ignore_prepared_cache_close_stmt=on;
    SET GLOBAL tidb_enable_inl_join_inner_multi_pattern=on;
    SET GLOBAL tidb_opt_derive_topn=on;
    SET GLOBALl tidb_opt_ordering_index_selectivity_threshold=0.0001;
    SET GLOBALl tidb_runtime_filter_mode=LOCAL;
    SET GLOBAL tidb_opt_enable_mpp_shared_cte_execution=on;
    SET GLOBAL tidb_rc_read_check_ts=on;
    SET GLOBAL tidb_guarantee_linearizability=off;
    SET GLOBAL tidb_enable_historical_stats=off;
    SET GLOBAL tidb_analyze_skip_column_types="json,blob,mediumblob,longblob,mediumtext,longtext";
    SET GLOBAL tidb_opt_prefer_range_scan=on;
    SET GLOBAL tidb_max_chunk_size=128;
    SET @@GLOBAL.tidb_enable_stmt_summary=0;
    SET GLOBAL tidb_enable_stmt_summary=off;
    SET GLOBAL tidb_enable_collect_execution_info=off;
    SET GLOBAL tidb_enable_resource_control=off;
    SET GLOBAL tidb_txn_assertion_level=off;
````

    ```

### Benchmark executor

The benchmark executor sends SQL queries to the TiDB cluster. In this test, its hardware configuration is as follows:

- Machine type: Amazon EC2 (us-west-2)
- Instance type: c6a.2xlarge

## Test steps

This section introduces how to perform the TPC-C performance test step by step.

1. In the [TiDB Cloud console](https://tidbcloud.com/), create a TiDB Cloud Dedicated cluster that meets the [test environment](#tidb-cluster) requirements.

    For more information, see [Create a TiDB Cloud Dedicated cluster](/tidb-cloud/create-tidb-cluster.md).

2. On the benchmark executor, connect to the newly created cluster and create a database named `tpcc`.

    To connect to the cluster, see [Connect to TiDB Cloud Dedicated via Private Endpoint](/tidb-cloud/set-up-private-endpoint-connections.md).

    To create the `tpcc` database, execute the following SQL statement:

    ```sql
    CREATE DATABASE tpcc;
    ```

3. Load TPC-C data to the `tpcc` database.

    1. The test in this document is implemented based on [go-tpc](https://github.com/pingcap/go-tpc). You can download the test program using the following command:

>        ```shell
>        curl --proto '=https' --tlsv1.2 -sSf https://raw.githubusercontent.com/pingcap/go-tpc/master/install.sh | sh
>        ```

Reviewer 2: ````suggestion
        ```shell
        curl --proto '=https' --tlsv1.2 -sSf https://raw.githubusercontent.com/pingcap/go-tpc/master/install.sh | sh
        ```
````

>     - The test in this document is implemented based on [go-tpc](https://github.com/pingcap/go-tpc). You can download the test program using the following command:
> 
>        ```shell
>        curl --proto '=https' --tlsv1.2 -sSf https://raw.githubusercontent.com/pingcap/go-tpc/master/install.sh | sh
>        ```
> 
>     - Run the following `go-tpc tpcc` command to import 900 warehouses to the `tpcc` database. Replace `${HOST}`, `${THREAD}`, and `${PASSWORD}` with your actual values. This document conducts three tests with the `${THREAD}` value of `50`, `100`, and `200`.

Reviewer: ```suggestion
    1. The test in this document is implemented based on [go-tpc](https://github.com/pingcap/go-tpc). You can download the test program using the following command:

       ```shell
       curl --proto '=https' --tlsv1.2 -sSf https://raw.githubusercontent.com/pingcap/go-tpc/master/install.sh | sh
       ```

    2. Run the following `go-tpc tpcc` command to import 900 warehouses to the `tpcc` database. Replace `${HOST}`, `${THREAD}`, and `${PASSWORD}` with your actual values. This document conducts three tests with the `${THREAD}` value of `50`, `100`, and `200`.
```

>     2. Run the following `go-tpc tpcc` command to import 900 warehouses to the `tpcc` database. Replace `${HOST}`, `${THREAD}`, and `${PASSWORD}` with your actual values. This document conducts three tests with the `${THREAD}` value of `50`, `100`, and `200`.

Reviewer 2: ```suggestion
    2. Run the following `go-tpc tpcc` command to import 900 warehouses to the `tpcc` database. Replace `${HOST}`, `${PORT}`, `${THREAD}`, and `${PASSWORD}` with your actual values. This document conducts three tests with the `${THREAD}` value of `50`, `100`, and `200`.
```

>        ```shell
>        go-tpc tpcc --host ${HOST} --warehouses 900 prepare -P 4000 -D tpcc -T ${THREAD} --time 2h0m0s -p ${PASSWORD} --ignore-error  --parts=12 --partition-type=4
>        ```

Reviewer 2: ````suggestion
        ```shell
        go-tpc tpcc --host ${HOST} --warehouses 900 prepare -P ${PORT} -D tpcc -T ${THREAD} --time 2h0m0s -p ${PASSWORD} --ignore-error --parts=12 --partition-type=4
        ```
````

4. To ensure that the TiDB optimizer can generate the optimal execution plan, execute the following SQL statements to collect statistics before conducting the TPC-C test:

    ```sql
    ANALYZE TABLE customer;
    ANALYZE TABLE district;
    ANALYZE TABLE history;
    ANALYZE TABLE item;
    ANALYZE TABLE new_order;
    ANALYZE TABLE order_line;
    ANALYZE TABLE orders;
    ANALYZE TABLE stock;
    ANALYZE TABLE warehouse;
    ```

    To accelerate the collection of statistics, execute the following SQL statements before collecting:

    ```sql
    SET tidb_build_stats_concurrency=16;
    SET tidb_distsql_scan_concurrency=16;
    SET tidb_index_serial_scan_concurrency=16;
    ```

5. Run the following command to warm up TPC-C performance tests on different workloads. Replace `${HOST}`, `${PORT}`, `${THREAD}`, and `${PASSWORD}` with your actual values.

    ```shell

>     go-tpc tpcc --host ${HOST} --warehouses 900 run -P ${HOST} -D tpcc -T ${THREAD} --time 2h0m0s -p ${PASSWORD} --ignore-error  --count=1000000

Reviewer 2: ```suggestion
    go-tpc tpcc --host ${HOST} --warehouses 900 run -P ${PORT} -D tpcc -T ${THREAD} --time 2h0m0s -p ${PASSWORD} --ignore-error  --count=1000000
```

    ```

6. Run the following `go-tpc tpcc` command to conduct stress tests on the TiDB Cloud Dedicated cluster.

    ```shell
    loads=("--threads=50 --count=1800000" "--threads=100 --count=2400000" "--threads=200 --count=2700000")
    for load in "${loads[@]}"; do
    sleep 60

>     go-tpc tpcc run --warehouses 900 --host {HOST} --db=tpcc $load -p ${PASSWORD} --ignore-error

Reviewer 2: ```suggestion
    go-tpc tpcc run --warehouses 900 --host ${HOST} -P ${PORT} --db=tpcc $load -p ${PASSWORD} --ignore-error
```
